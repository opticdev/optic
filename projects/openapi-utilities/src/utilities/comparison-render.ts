import { ResultWithSourcemap, IChange, OperationLocation } from '..';

import groupBy from 'lodash.groupby';
import isUrl from 'is-url';
import { Instance as Chalk } from 'chalk';
import { getOperationsModifsLabel } from './count-changed-operations';

const getIndent = (depth: number): string => ' '.repeat(depth * 2);

// raw string value
const formatRawValue = (value: string, indent: string): string => {
  try {
    const parsedValue = JSON.parse(value);
    return (
      indent +
      JSON.stringify(parsedValue, null, 2).replace(/\n/g, '\n' + indent)
    );
  } catch (e) {
    return value;
  }
};

function* generateComparisonLogs(
  comparison: {
    results: ResultWithSourcemap[];
    changes: IChange[];
  },
  options: {
    output: 'pretty' | 'plain';
    verbose: boolean;
  }
) {
  const chalk = new Chalk({ level: options.output === 'plain' ? 0 : 1 });
  const totalNumberOfChecks = comparison.results.length;
  const failedNumberOfChecks = comparison.results.filter(
    (result) => !result.passed && !result.exempted
  ).length;
  const exemptedFailedNumberOfChecks = comparison.results.filter(
    (result) => !result.passed && result.exempted
  ).length;
  const passedNumberOfChecks = totalNumberOfChecks - failedNumberOfChecks;
  const operationsModifsLabel = getOperationsModifsLabel(comparison.changes);
  const groupedResults = groupBy(
    comparison.results,
    (result) =>
      `${
        (result.change.location.conceptualLocation as OperationLocation).method
      }-${
        (result.change.location.conceptualLocation as OperationLocation).path
      }`
  );

  for (const operationResults of Object.values(groupedResults)) {
    const conceptualLocation =
      operationResults[0].change.location.conceptualLocation;

    const allPassed = operationResults.every(
      (result) => result.passed || result.exempted
    );
    const renderedResults = operationResults.filter(
      (result) => options.verbose || (!result.passed && !result.exempted)
    );
    const resultNode = allPassed
      ? chalk.bold.bgGreen.white(' PASS ')
      : chalk.bold.bgRed.white(' FAIL ');

    if (!('path' in conceptualLocation)) {
      yield `${getIndent(1)}${resultNode} ${chalk.bold('Specification')}`;
    } else {
      const { method, path } = conceptualLocation;
      yield;
      `${getIndent(1)}${resultNode} ${chalk.bold(
        method.toUpperCase()
      )} ${path}`;
    }

    for (const result of renderedResults) {
      const icon = result.passed
        ? chalk.green('✔')
        : result.exempted
        ? chalk.white('✔')
        : chalk.red('x');

      const rulePrefix = result.type ? `${result.type} rule` : 'rule';
      yield `${getIndent(2)}${rulePrefix}: ${result.name ?? ''}${
        result.exempted ? ' (exempted)' : ''
      }`;

      if (!result.passed && !result.exempted) {
        yield getIndent(3) + chalk.red(`${icon} ${result.error}`);
        if (result.expected && result.received) {
          yield getIndent(3) + chalk.red('Expected Value:');
          yield formatRawValue(result.expected, getIndent(3));
          yield getIndent(3) + chalk.red('Received Value:');
          yield formatRawValue(result.received, getIndent(3));
        }
      }
      if (!result.passed && result.exempted) {
        yield getIndent(3) + `${icon} ${result.error}`;
        if (result.expected && result.received) {
          yield getIndent(3) + 'Expected Value:';
          yield formatRawValue(result.expected, getIndent(3));
          yield getIndent(3) + 'Received Value:';
          yield formatRawValue(result.received, getIndent(3));
        }
      }

      if (result.docsLink) {
        yield `${getIndent(3)}Read more in our API Guide (${result.docsLink})`;
      }
      if (result.sourcemap) {
        yield `${getIndent(3)}at ${
          isUrl(result.sourcemap.filePath)
            ? `${chalk.underline(result.sourcemap.filePath)} line ${
                result.sourcemap.startLine
              }`
            : chalk.underline(
                `${result.sourcemap.filePath}:${result.sourcemap.startLine}:${result.sourcemap.startPosition}`
              )
        }`;
      }
      yield '';
    }
    yield '\n';
  }

  yield operationsModifsLabel;
  yield chalk.green.bold(`${passedNumberOfChecks} checks passed`);
  yield chalk.red.bold(`${failedNumberOfChecks} checks failed`);

  if (exemptedFailedNumberOfChecks > 0) {
    yield chalk.bold(`${exemptedFailedNumberOfChecks} checks exempted`);
  }
}

export const getComparisonLogs = (
  comparison: {
    results: ResultWithSourcemap[];
    changes: IChange[];
  },
  options: {
    output: 'pretty' | 'plain';
    verbose: boolean;
  }
) => [...generateComparisonLogs(comparison, options)];

export const logComparison = (
  comparison: {
    results: ResultWithSourcemap[];
    changes: IChange[];
  },
  options: {
    output: 'pretty' | 'plain';
    verbose: boolean;
  }
) => {
  for (const log of generateComparisonLogs(comparison, options)) {
    console.log(log);
  }
};
