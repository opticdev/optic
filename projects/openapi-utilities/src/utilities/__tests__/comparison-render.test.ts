import { getComparisonLogs } from '../comparison-render';

const comparisonWithFailedResult = {
  results: [
    {
      passed: false,
      condition: 'have snake case keys',
      where: 'added field: strangeness',
      isMust: true,
      isShould: false,
      change: {
        location: {
          jsonPath:
            '/paths/~1orgs~1{org_id}~1thing~1{thing_id}/get/responses/200/content/application~1vnd.api+json/schema/properties/data/properties/attributes/properties/strangeness',
          conceptualPath: [
            'operations',
            '/orgs/{}/thing/{}',
            'get',
            'responses',
            '200',
            'application/vnd.api+json',
            'data',
            'attributes',
            'strangeness',
          ],
          kind: 'field',
          conceptualLocation: {
            method: 'get',
            path: '/orgs/{org_id}/thing/{thing_id}',
            inResponse: {
              statusCode: '200',
              body: {
                contentType: 'application/vnd.api+json',
              },
            },
            jsonSchemaTrail: ['data', 'attributes', 'strangeness'],
          },
        },
        added: {
          key: 'strangeness',
          flatSchema: {
            type: 'number',
            description:
              'The amount of strangeness this thing adds or removes from the situation.',
            example: 42,
          },
          required: true,
        },
        changeType: 'added',
      },
      sourcemap: {
        filePath:
          '/end-end-tests/api-standards/resources/thing/2021-11-10/001-ok-add-property-field.yaml',
        startLine: 221,
        endLine: 224,
        preview: '',
        startPosition: 10080,
        endPosition: 10233,
      },
    },
  ],
  changes: [
    {
      location: {
        jsonPath: '/paths/~1user~1{username}/get',
        conceptualPath: ['operations', '/user/{}', 'get'],
        kind: 'operation',
        conceptualLocation: { method: 'get', path: '/user/{username}' },
      },
      added: {
        tags: ['user'],
        summary: 'Get user by user name',
        operationId: 'getUserByName',
        method: 'get',
        pathPattern: '/user/{username}',
      },
      changeType: 'added',
    },
  ],
} as any;

describe('getComparisonLogs', () => {
  test('pretty non verbose', () => {
    const res = getComparisonLogs(comparisonWithFailedResult, {
      output: 'pretty',
      verbose: false,
    });
    expect(res).toMatchSnapshot();
  });
  test('md non verbose', () => {
    const res = getComparisonLogs(comparisonWithFailedResult, {
      output: 'md',
      verbose: false,
    });
    expect(res).toMatchSnapshot();
  });
});
