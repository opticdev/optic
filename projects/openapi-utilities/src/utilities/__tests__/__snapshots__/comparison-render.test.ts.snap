// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComparisonLogs md non verbose 1`] = `
[
  "**FAIL** **GET** /orgs/{org_id}/thing/{thing_id}",
  "- rule: ",
  ":x: undefined",
  "at /end-end-tests/api-standards/resources/thing/2021-11-10/001-ok-add-property-field.yaml:221:10080",
  "",
  "
",
  "1 operation added",
  "**0 checks passed**",
  "**1 checks failed**",
]
`;

exports[`getComparisonLogs pretty non verbose 1`] = `
[
  "  [1m[41m[37m FAIL [39m[49m[22m [1mGET[22m /orgs/{org_id}/thing/{thing_id}",
  "    rule: ",
  "      [31m[31mx[39m[31m undefined[39m",
  "      at [4m/end-end-tests/api-standards/resources/thing/2021-11-10/001-ok-add-property-field.yaml:221:10080[24m",
  "",
  "
",
  "1 operation added",
  "[32m[1m0 checks passed[22m[39m",
  "[31m[1m1 checks failed[22m[39m",
]
`;
